Reporter options to truncate message and suppress stacktrace

james-stocks opened the pull request:
Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
(WIP) Reporter options to truncate message and suppress stacktrace

TODO: Option names are not JSON specific but only json reporter has been implemented.

Signed-off-by: James Stocks <jstocks@chef.io>

james-stocks commented on lib/inspec/runner.rb:
```diff
@@ -134,7 +134,18 @@ def render_output(run_data)
 
       @conf["reporter"].each do |reporter|
-        result = Inspec::Reporters.render(reporter, run_data)
+        reporter_opts = {}
```
Optionally the `run_data` could be directly modified here instead of passing options to the reporters. But I think there should be a distinction between the data that actually happened and the data that's reported

Schwad commented on lib/inspec/reporters.rb:
```diff
@@ -9,26 +9,26 @@
 module Inspec::Reporters
   # rubocop:disable Metrics/CyclomaticComplexity
-  def self.render(reporter, run_data)
+  def self.render(reporter, run_data, opts = {})
```
I think what's interesting here, is wouldn't this setup give us the power to use options on any of our reporters in the future? (With the first two being these specific ones relating to JSON)

Though I do see where we could be increasing data coupling (passing data to the reporters) -  it is with a single options hash that opens up these future feature possibilities.

james-stocks replied:
At this point, the `reporter` passed to this method is `["json", {"stdout"=>true}]`
That might be an existing hash that we can and should use to pass options for the reporter

james-stocks commented on lib/inspec/runner.rb:
```diff
@@ -134,7 +134,18 @@ def render_output(run_data)
+        if @conf.key?(:reporter_message_truncation)
+          reporter_opts[:message_truncation] = @conf['reporter_message_truncation']
+        elsif reporter[0] == 'json-automate'
+          reporter_opts[:message_truncation] = '200'
```
From chatting to @kekaichinose the default for Automate should maybe be to not truncate. On the principle that for compliance, all data should be reported. 
Then Automate users will need to opt in for the reduced report size.

Commit:
Move reporter option handling to base

They don't need to be json specific, they should apply to any reporter if the user has chosen these settings.

Signed-off-by: James Stocks <jstocks@chef.io>

clintoncwolfe commented on lib/inspec/runner.rb:
```diff
@@ -134,6 +134,8 @@ def render_output(run_data)
       return if @conf["reporter"].nil?
 
       @conf["reporter"].each do |reporter|
+        reporter[1][:include_backtrace] = @conf["reporter_backtrace_inclusion"] if @conf.key?(:reporter_backtrace_inclusion)
```
This seems odd on first reading - check for a Symbol key, then set based on a String key?

clintoncwolfe commented on lib/inspec/runner.rb:
```diff
@@ -134,6 +134,8 @@ def render_output(run_data)
       return if @conf["reporter"].nil?
 
       @conf["reporter"].each do |reporter|
+        reporter[1][:include_backtrace] = @conf["reporter_backtrace_inclusion"] if @conf.key?(:reporter_backtrace_inclusion)
```
These options are global to the run, not per-reporter. Inspec::Config (that is, `@conf`) is global, too - no need to make changes to `runner.rb` at all. You could just call it directly in `reporters/base.rb`

james-stocks replied:
I pushed another commit that leaves runner.rb unchanged

clintoncwolfe commented on lib/inspec/reporters/base.rb:
```diff
@@ -5,9 +5,36 @@ class Base
 
+    # Apply options such as message truncation and removal of backtraces
+    def apply_options_to_run_data
+      message_truncation = @config[:message_truncation] || "ALL"
```
Right here, you could call `runtime_config = Inspec::Config.cached` and get an instance of  the runtime config with access to the CLI options or JSON config - no need to sneak it into the per-reporter configs.

james-stocks replied:
Thanks, that sounds good! I think that solves the problem Alex mentioned where these options are not working from STDIN config.
It would make it a little odd that the class is instantiated with a `config` hash but we don't use that for config - but it does mean I can revert my changes to the runner.

clintoncwolfe commented on lib/inspec/reporters/base.rb:
```diff
@@ -5,9 +5,36 @@ class Base
+      trunc = -1
+      if message_truncation != "ALL"
+        if message_truncation.to_i.to_s != message_truncation
+          Inspec::Log.warn("Messages will not be truncated because #{message_truncation} is not an integer value")
```
Consider making this an error - if they provided the option but it is unusable, halt and show usage of the bad option is a reasonable thing to do. Continuing and generating too much data might be worse.

james-stocks replied:
My thinking for making this a warning is that:
 * The CLI, not this class, is responsible for config validation (even though it's not doing any currently...). 
 * At this point, the run has completed. It possibly took a long time and burned through billable compute resource. The reporter should generally never halt.

Possibly I could make the option strictly numeric and throw away the `ALL` special case - I did it this way so that those using InSpec in CI could have a variable driven run command where `message-truncation` is always set at the CLI and the value is defined by a CI variable.

clintoncwolfe replied:
Good point about the validation happening post-run - perhaps the validation should be moved to be here - https://github.com/inspec/inspec/blob/2e887a94afcca819da781d4774aa2a5a0b56785e/lib/inspec/config.rb#L328

james-stocks replied:
Thanks, that looks like the right spot for validation.
This warning was meant to supplementary to proper validation - but I think if there's validation in `config.rb` then this class also doing a check is overkill - so I'll just remove this warning.

clintoncwolfe commented on lib/inspec/reporters/base.rb:
```diff
@@ -5,9 +5,36 @@ class Base
     def initialize(config)
       @config = config
       @run_data = config[:run_data]
+      apply_options_to_run_data
```
I know naming things is hard and no one will ever agree, but perhaps "apply_message_size_limits"? If there are future options to implement, they would go in their own methods.

james-stocks replied:
I was foreseeing the opposite - that this method _would_ handle any future options. I can rename the method like you suggest so that it is named specific to the thing it _currently_ does, and we can discuss handling future options when we get to that point.

clintoncwolfe reviewed (commented):
The algorithm seems fine, main thing I note is there is no need stash the CLI options on the per-reporter configs. Few other suggestions, most minor.

Commit:
Do not inject config to base reporter

The base reporter class should access the InSpec config cache, instead of the runner needing to inject config into the base reporter

Signed-off-by: James Stocks <jstocks@chef.io>

clintoncwolfe reviewed (approved):
This looks good, thanks!

Schwad commented on lib/inspec/reporters/base.rb:
```diff
@@ -5,9 +5,31 @@ class Base
 
+    # Apply options such as message truncation and removal of backtraces
+    def apply_report_resize_options
+      runtime_config = Inspec::Config.cached.respond_to?(:final_options) ? Inspec::Config.cached.final_options : {}
```
This is good to know for future

Schwad commented on lib/inspec/reporters/base.rb:
```diff
@@ -5,9 +5,31 @@ class Base
+      trunc = message_truncation == "ALL" ? -1 : message_truncation.to_i
+      include_backtrace = runtime_config[:reporter_backtrace_inclusion].nil? ? true : runtime_config[:reporter_backtrace_inclusion]
+
+      @run_data[:profiles]&.each do |p|
```
I had a quick look with `benchmark/ips` in the Terminal with an identical data structure to see if I could find another way to make this change without the loops but as far as I could tell this is the best approach!

james-stocks replied:
Yes I don't much like these nested loops.
Note the cautious  `&.` because at each level the report might be missing any occurrences of the nested elements.

Schwad reviewed (approved):
Looking great, thanks! 👍

james-stocks commented on lib/inspec/config.rb:
```diff
@@ -360,6 +360,12 @@ def validate_reporters!(reporters)
       raise ArgumentError, "The option --reporter can only have a single report outputting to stdout." if stdout_reporters > 1
+
+      if (truncation = @merged_options["reporter_message_truncation"])
+        unless truncation == "ALL" || truncation.to_i.to_s == truncation
```
This is wrong if the value is already an Integer. Usually it will be a string (via the CLI) but can be an Integer via JSON config.
@clintoncwolfe 's PR for typing inputs will also cause this to be wrong, as that PR causes inputs to be typed where possible

clintoncwolfe replied:
#5011 will only impact the `--input` CLI option

james-stocks replied:
D'oh, good point!

Schwad reviewed (changes requested):
Look to handle actual integer via JSON:

````
echo '{"reporter_backtrace_inclusion": false, "reporter_message_truncation": 100, "reporter":["json-automate"]}' | be inspec exec ~/git/mycompliance-profile/myprofile1 --json-config=-
#=> reporter_message_truncation is set to 100. It must be set to an integer value or ALL to indicate no truncation.
```

alexpop reviewed (approved):
Cheers 🚀 🚀 🚀
